### ISSUE TITLE
prefer-const considers correct code to be incorrect

### ISSUE THREAD

**Issue by NickHeiner** [eslint/eslint#11055]
**What rule do you want to change?**
`prefer-const`

**Does this change cause the rule to produce more or fewer warnings?**
Fewer

**How will the change be implemented? (New option, new default behavior, etc.)?**
For backwards compat, this change may need to be behind a new option. But I think the more correct thing to do to change the default behavior.

**Please provide some example code that this change will affect:**

```js
let id;
const res = doAction(() => {
    console.log(id);
});
id = res.getId();
```

**What does the rule currently do for this code?**
Flags an error: `'id' is never reassigned. Use 'const' instead`

**What will the rule do after it's changed?**
I don't think that this code should be considered invalid. This is a valid situation where I'd want to use `let`. It's true that `id` is never re-assigned. However, this is a not-uncommon code pattern where a variable is read lexically before, but temporally after, it's written. 

**Are you willing to submit a pull request to implement this change?**
Maybe, but realistically it would probably take me a while to get to it.

**Comment by ljharb** [comment]
Why is that code pattern better than:
```js
const id = res.getId();
const res = doAction(() => {
    console.log(id);
});
```
?

If the `doAction` callback is asynchronous, then this would work too:
```js
const res = doAction(() => {
    console.log(id);
});
const id = res.getId();
```

**Comment by mysticatea** [comment]
What is different to [ignoreReadBeforeAssign](https://eslint.org/docs/rules/prefer-const#options) option?

**Comment by ljharb** [comment]
@NickHeiner lol ok, fair enough :-) i still don't understand why you'd need an undefined variable in there tho, since the second example works too if the callback is ran later.

**Comment by NickHeiner** [comment]
`ignoreReadBeforeAssign`, in conjunction with @ljharb's second example, is what I needed. Thanks! Sorry for opening this issue before being thorough enough with the docs. :smile:
